src/bcounter_mgr.erl

```diff
@@ -31,201 +31,294 @@
 
-handle_cast({transfer, {Key, Amount, Requester}}, #state{last_transfers=LT}=State) ->
-    NewLT = cancel_consecutive_req(LT, ?GRACE_PERIOD),
-    MyDCId = dc_utilities:get_my_dc_id(),
-    case can_process(Key, Requester, NewLT) of
-        true ->
-            {SKey, Bucket} = Key,
-            BObj = {SKey, ?DATA_TYPE, Bucket},
-            % try to transfer locks, might return {error,no_permissions} if not enough permissions are available locally
-            _ = antidote:update_objects(ignore, [], [{BObj, transfer, {Amount, Requester, MyDCId}}]),
-            {noreply, State#state{last_transfers=orddict:store({Key, Requester}, erlang:timestamp(), NewLT)}};
-        _ ->
-            {noreply, State#state{last_transfers=NewLT}}
-    end.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
 
-handle_call({consume, Key, {Op, {Amount, _}}, BCounter}, _From, #state{req_queue=RQ}=State) ->
-    MyDCId = dc_utilities:get_my_dc_id(),
-    case ?DATA_TYPE:generate_downstream_check({Op, Amount}, MyDCId, BCounter, Amount) of
-        {error, no_permissions} = FailedResult ->
-            Available = ?DATA_TYPE:localPermissions(MyDCId, BCounter),
-            UpdtQueue=queue_request(Key, Amount - Available, RQ),
-            {reply, FailedResult, State#state{req_queue=UpdtQueue}};
-        Result ->
-            {reply, Result, State}
-    end.
+init([]) ->
+    Timer = erlang:send_after(?TRANSFER_FREQ, self(), transfer_periodic),
+    {ok, #state{transfer_timer = Timer}}.
+
+handle_call(Request = {_, {decrement, {_, _}}, _}, _From, State = #state{pending_transfer_requests = PendingTransferRequests}) ->
+    {Result, NewPendingTransferRequests} = decrement(Request, PendingTransferRequests),
+    {reply, Result, State#state{pending_transfer_requests = NewPendingTransferRequests}};
+
+handle_call(get_pending_transfer_requests, _From, State = #state{pending_transfer_requests = PendingTransferRequests}) ->
+    {reply, PendingTransferRequests, State};
+
+handle_call({set_transfer_periodic_active, Active}, _From, State) ->
+    {reply, ok, State#state{transfer_periodic_active = Active}}.
+
+handle_cast({transfer, Request}, State = #state{last_transfers = LastTransfers}) ->
+    NewLastTransfers = transfer(Request, LastTransfers),
+    {noreply, State#state{last_transfers = NewLastTransfers}}.
 
-handle_info(transfer_periodic, #state{req_queue=RQ0, transfer_timer=OldTimer}=State) ->
+handle_info(transfer_periodic, State = #state{transfer_periodic_active = false}) -> {noreply, restart_timer(State)};
+handle_info(transfer_periodic, State = #state{pending_transfer_requests = PendingTransferRequests}) ->
+    NewPendingTransferRequests = transfer_periodic(PendingTransferRequests),
+    {noreply, restart_timer(State#state{pending_transfer_requests = NewPendingTransferRequests})}.
+
+terminate(_Reason, #state{transfer_timer = OldTimer}) ->
     _ = erlang:cancel_timer(OldTimer),
-    RQ = clear_pending_req(RQ0, ?REQUEST_TIMEOUT),
```

Comment:
Why is this simplification valid?

Reply:
It was replaced by 
`ClearedPendingTransferRequests = clear_old_transfer_requests(PendingTransferRequests, ?REQUEST_TIMEOUT),`
in line 214
I moved the code a bit so that the gen_server callbacks don't contain the logic (this can be reverted)